internal/consensus/state.go

```diff
@@ -1508,6 +1508,19 @@ func (cs *State) defaultDoPrevote(ctx context.Context, height int64, round int32
 		}
 	}
 
+	stateMachineValidBlock, err := cs.blockExec.ProcessProposal(cs.ProposalBlock)
+	if err != nil {
+		logger.Error("state machine returned an error when trying to process proposal block", "err", err)
```

Comment:
Not sure if it makes sense to make this change now or in later fixups, but we should issue the prevote for nil in this error condition check.

Reply:
AFAIU, this is an unexpected error from ABCI infra, or maybe from the App (the `nil`-prevoting seems to take place in line 1520 below).
In this case I'd propose we do the same we do for existing API calls (e.g., `BeginBLock`, `DeliverTx`, `EndBlock`)... I'll look that up and then we can see what to do here

Reply:
Took a closer look at the code and actually in this "if" we fall through to the next one. When `BlockExecutor`'s method `ProcessProposal` returns an error, it also returns `false` as first parameter, so the condition at the next "if" will be true and we'll prevote `nil`.
Maybe adding `stateMachineValidBlock = false` inside this "if" will help readability. What do you think?

Reply:
Ah, thanks for diving in. This feels like a bit of a conundrum since we didn't have this kind of application logic anywhere in the consensus engine at all before. The error type returned by `ProcessProposal` is an `ErrInvalidBlock`, which _sounds_ like the block should not be considered valid, however, the `IsOk` check that happens on the response object appears to cover that condition instead. This means that this is a check is determining if communication with the app was successful or not. 

In this situation, I think we get to choose how Tendermint responds when it has an issue communicating with the application. Our options are 1. Issue a prevote for nil 2. Crash, 3. Return from the method, issuing no prevote.

Thinking this through a bit, I would actually be in favor of _crashing_ since it's not clear how to recover from this condition. The application is effectively not participating in consensus. We could hide the fact that it's not participating in consensus by just logging the error and continuing. I'd definitely be interested to hear what other people have to say about this though.

Reply:
I tend to agree with you: as we don't know how to react to this, I'd crash. If we try to (keep calm and) carry on, we might end up with another _weird_ "manifestation" of the problem later on, which could be a nightmare to diagnose. In these cases, as a troubleshooter, I prefer to see where the problem actually started, in the form of a backtrace (or core file, or similar) rather than having to trace the problem back (via logs, etc).

But you're right: let's see if folks think the same way or have different views.

Reply:
I think I was involved in a similar discussion before where we had these three options. My ideas then was to log the error and not vote at all. My thinking was that crashing seems a bit extreme if there was only something temporarily wrong with for example the socket connection but was eventually able to recover in the following round.

As far as other nodes are concerned, not voting looks the same as crashing i.e. it's an omission fault. Then again I don't probably have sufficient knowledge on how the connection could fail and am not entirely opposed to doing that.

Reply:
We can further discuss about this if you have concerns. Mine is that If the communication is disrupted between Tendermint and the App (e.g., when using a socket), that is something we don't know how to recover from, even at a theoretical level. For instance, what if the error at this point is produced because the App just missed the last `FinalizeBlock` message?

I propose to leave the `panic`. If that gets in the way of some user in an unreasonable way, we can then "learn the lesson" and change it to an omission fault.